services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/AttributeName.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+/**
+ * High-level representation of a DynamoDB 'attribute name' that can be used in various situations where the API requires
+ * or accepts an Attrbute Name. In addition various convenience methods are provided to help manipulate expressions.
+ * <p>
+ * At a minimum, an AttributeName must contain a string that represents the name of the Attribute.
+ * <p>
+ * Optionally, nested attribute names can be added with tokens using the addNestedAttributeName .
+ * Every call to addNestedAttributeName represent next level of Nesting Attribute.
+ * If we want to denote three level of Nesting attribute name then addNestedAttributeName should be called thrice.
+ *
+ * <p>
+ * Example, if we want to represent a two level nested attribute outerAttribute.innerAttributeLevel1.innerAttributeLevel2
+ * then it can be represented as :-
+ * {@code
+ * AttributeName myAttributeName = AttributeName.builder()
+ * .name("outerAttribute).addNested("innerAttributeLevel1).nestedAttributeName("innerAttributeLevel2").build();
+ * }
+ */
+@SdkPublicApi
+public final class AttributeName {
+
+    private final String name;
+    private final List<String> nestedAttributeNames;
```

Comment:
Ignoring the builder entirely, I think we should store this just as a single list. If it's a single element, then it's not nested, if there are multiple elements then it's nested.